File: typescript/src/ray.ts
import { Vec3 } from './vec3'
import { Model, Hittable } from './model'

export interface HitRay {
  ray: Ray
  point: Vec3
  hittable: Hittable
  distance: number
}

export interface HitModel extends HitRay {
  model: Model
}

export class Ray {
  constructor(public start: Vec3, public direction: Vec3) {}

  getPoint(distance: number): Vec3 {
    return this.start.add(this.direction.scale(distance))
  }
}

File: rust/src/ray.rs
use nalgebra::Point3;

use crate::model::Vec3;
use crate::world::World;

pub struct Ray {
    pub origin: Point3<f64>,
    pub dir: Vec3,
}

impl Ray {
    pub fn new(origin: Point3<f64>, dir: Vec3) -> Ray {
        Ray { origin, dir }
    }

    pub fn get_point(&self, distance: f64) -> Point3<f64> {
        self.origin + self.dir * distance
    }
}

pub struct HitRecord {
    pub dist: f64,
    pub hit_point: Point3<f64>,
    pub normal: Vec3,
    pub material_id: usize,
}

pub struct RayHit<'a> {
    pub ray: &'a Ray,
    pub hit_point: Point3<f64>,
    pub normal: Vec3,
    pub world: &'a World,
    pub depth: i32,
    pub material_id: usize,
}
